reference/best-practices: add PD scheduling best practice (#1655)
pingcap/docs: dev/glossary.md

An Operator can be computed and generated by a Scheduler, or created by an external API.

### Operator Step

An Operator Step is a step in the execution of an Operator. An Operator normally contains multiple Operator steps.

Currently, available Steps generated by PD include:

- `TransferLeader`: migrate a Region Leader to a specified Peer
- `AddPeer`: add Followers to a specified Store
- `RemovePeer`: delete a Region Peer
- `AddLearner`: add a Region Learner to a specified Store
- `PromoteLearner`: promote a specified Learner to a voting member
- `SplitRegion`: split a Region in two

## P

### `Pending`/`Down`

`Pending` and `Down` are two special states of Peer. `Pending` indicates that the raft log of Followers or Learners is vastly different from that of Leader, and Followers in `Pending` cannot be elected as Leader. `Down` refers to a state that a Peer ceases to respond to the corresponding Leader for a long time, which usually means that the corresponding node is down or isolated from the network.

## R

### Region/Peer/Raft Group

> Each Region maintains a continuous piece of data for the cluster (an average of about 96 MiB by default), each of which stores multiple replicas in different Stores (3 replicas by default) and each replica is referred as a Peer. Multiple Peers of the same Region synchronize data via raft protocol, so Peers also refer to members of a raft instance. TiKV uses the multi-raft pattern to manage data, that is, each Region has a corresponding, standalone raft instance (also known as a Raft Group).

Reviewer: ```suggestion
Each Region maintains a continuous piece of data for the cluster (an average of about 96 MiB by default), each of which stores multiple replicas in different Stores (3 replicas by default) and each replica is referred as a Peer. Multiple Peers of the same Region synchronize data via raft protocol, so Peers also refer to members of a raft instance. TiKV uses the multi-raft pattern to manage data, that is, each Region has a corresponding, isolated Raft Group.
```

So folks don't assume it's a single node raft group :)

### Region Split

Regions in the TiKV cluster are gradually split and generated as data write continues. The process of splitting is called Region Split.

> The mechanism of Region Split is to build an initial Region to cover the entire key space in cluster initialization, and then generate a new Region through Split every time the Region data reaches a certain amount.

Reviewer 2: ```suggestion
The mechanism of Region Split is to build an initial Region to cover the entire key space in cluster initialization, and then generate new Regions through splitting existing ones every time the Region data reaches a certain amount.
```

## S

### Scheduler

> Scheduler is a component in PD that generates scheduling tasks. Each scheduler in PD runs independently and serves different purposes. Common schedulers and their purposes are:

Reviewer: ```suggestion
Schedulers are a component in PD that generates scheduling tasks. Each scheduler in PD runs independently and serves different purposes. Common schedulers and their purposes are:
```

- `balance-leader-scheduler`: maintains Leader balance of different nodes
- `balance-region-scheduler`: maintains Peer balance of different nodes
- `hot-region-scheduler`: maintains hot Region balance of different nodes
- `evict-leader-{store-id}`: evicts all Leaders of a node (often used for rolling upgrades)

### Store

> A Store in PD refers to the storage node in the cluster (an instance of `tikv-server`). Each Store has a corresponding TiKV instance. This means if multiple TiKV instances are deployed on the same host or even on the same disk, these instances still correspond to different Stores.

Reviewer 2: ```suggestion
A Store refers to the storage node in the TiKV cluster (an instance of `tikv-server`). Each Store has a corresponding TiKV instance.
```

> A Store in PD refers to the storage node in the cluster, also a instance of tikv-server. Each Store has a corresponding TiKV instance, which means if multiple TiKV instances are deployed on the same host or even on the same disk, these instances still correspond to different Stores.

Reviewer: ```suggestion
A Store in PD refers to the storage node in the cluster (an instance of `tikv-server`). Each Store has a corresponding TiKV instance, which means if multiple TiKV instances are deployed on the same host or even on the same disk, these instances still correspond to different Stores.
```
